conans/client/conan_api.py

```diff
@@ -1249,53 +1272,90 @@ def update_lock(self, old_lockfile, new_lockfile, cwd=None):
+
+        if path:
+            ref_or_path = _make_abs_path(path, cwd)
+            if not os.path.isfile(ref_or_path):
+                raise ConanException("Conanfile does not exist in %s" % ref_or_path)
+        else:  # reference
+            ref_or_path = ConanFileReference.loads(reference)
+
+        phost = pbuild = graph_lock = None
+        if lockfile:
+            lockfile = _make_abs_path(lockfile, cwd)
+            graph_lock_file = GraphLockFile.load(lockfile, self.app.cache.config.revisions_enabled)
+            phost = graph_lock_file.profile_host
+            pbuild = graph_lock_file.profile_build
+            graph_lock = graph_lock_file.graph_lock
+            graph_lock.relax()
+
+        if not phost:
+            phost = profile_from_args(profile_host.profiles, profile_host.settings,
+                                      profile_host.options, profile_host.env, cwd, self.app.cache)
+
+        if profile_build and not pbuild:
+            # Only work on the profile_build if something is provided
+            pbuild = profile_from_args(profile_build.profiles, profile_build.settings,
+                                       profile_build.options, profile_build.env, cwd, self.app.cache)
+
+        root_ref = ConanFileReference(name, version, user, channel, validate=False)
+        phost.process_settings(self.app.cache)
+        if pbuild:
+            pbuild.process_settings(self.app.cache)
+        graph_info = GraphInfo(profile_host=phost, profile_build=pbuild, root_ref=root_ref)
+        graph_info.graph_lock = graph_lock
+
         recorder = ActionRecorder()
         # FIXME: Using update as check_update?
         remotes = self.app.load_remotes(remote_name=remote_name, check_updates=update)
-        deps_graph = self.app.graph_manager.load_graph(reference, None, graph_info, build, update,
-                                                       False, remotes, recorder)
-
+        deps_graph = self.app.graph_manager.load_graph(ref_or_path, None, graph_info, build, update,
+                                                       update, remotes, recorder)
         print_graph(deps_graph, self.app.out)
-        lockfile = _make_abs_path(lockfile)
-        graph_info.save_lock(lockfile)
-        self.app.out.info("Generated lockfile")
+
+        # The computed graph-lock by the graph expansion
+        graph_lock = graph_info.graph_lock
+        # Pure graph_lock, no more graph_info mess
+        graph_lock_file = GraphLockFile(phost, pbuild, graph_lock)
+        if lockfile:
+            new_graph_lock = GraphLock(deps_graph, self.app.config.revisions_enabled)
+            # check if the lockfile provided was used or not
+            new_graph_lock.check_contained(graph_lock)
+            graph_lock_file = GraphLockFile(phost, pbuild, new_graph_lock)
+        if base:
+            graph_lock_file.only_recipes()
+
+        lockfile_out = _make_abs_path(lockfile_out or "conan.lock")
```

Comment:
the argument is required

```suggestion
        lockfile_out = _make_abs_path(lockfile_out)
```